From 1f8cc00292f7977ef963fcf3036db608d622f671 Mon Sep 17 00:00:00 2001
From: Alvin Chang <alvinga@andestech.com>
Date: Fri, 8 Dec 2023 08:48:43 +0800
Subject: [PATCH 134/241] target/riscv: Fix tcontrol fields changed when
 trapping to M-mode

The tcontrol in debug Sdtrig specification says:
1. When any trap into M-mode is taken, mpte is set to the value of mte.
2. When any trap into M-mode is taken, mte is set to 0.

Therefore, the tcontrol bit fields should only be changed when trapping
to M-mode. Fix it since last commit we change it when trapping to
S-mode.

Signed-off-by: Alvin Chang <alvinga@andestech.com>
---
 target/riscv/cpu_helper.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/target/riscv/cpu_helper.c b/target/riscv/cpu_helper.c
index c7b9e0bc0e..b8f1197620 100644
--- a/target/riscv/cpu_helper.c
+++ b/target/riscv/cpu_helper.c
@@ -1788,13 +1788,6 @@ void riscv_cpu_do_interrupt(CPUState *cs)
             env->hstatus = set_field(env->hstatus, HSTATUS_GVA, write_gva);
         }
 
-        if (riscv_cpu_cfg(env)->ext_sdtrig_tcontrol) {
-            s = env->tcontrol;
-            s = set_field(s, TCONTROL_MPTE, get_field(s, TCONTROL_MTE));
-            s = set_field(s, TCONTROL_MTE, 0);
-            env->tcontrol = s;
-        }
-
         s = env->mstatus;
         s = set_field(s, MSTATUS_SPIE, get_field(s, MSTATUS_SIE));
         s = set_field(s, MSTATUS_SPP, env->priv);
@@ -1826,6 +1819,14 @@ void riscv_cpu_do_interrupt(CPUState *cs)
             riscv_cpu_set_virt_enabled(env, 0);
         }
 
+        /* Trapping to M-mode, set tcontrol in debug Sdtrig extension. */
+        if (riscv_cpu_cfg(env)->ext_sdtrig_tcontrol) {
+            s = env->tcontrol;
+            s = set_field(s, TCONTROL_MPTE, get_field(s, TCONTROL_MTE));
+            s = set_field(s, TCONTROL_MTE, 0);
+            env->tcontrol = s;
+        }
+
         s = env->mstatus;
         s = set_field(s, MSTATUS_MPIE, get_field(s, MSTATUS_MIE));
         s = set_field(s, MSTATUS_MPP, env->priv);
-- 
2.34.1

